diff --git a/libril/ril.cpp b/libril/ril.cpp
index d7744c0..561ba55 100644
--- a/libril/ril.cpp
+++ b/libril/ril.cpp
@@ -3804,6 +3804,7 @@ static void responseSimStatusV5(Parcel &p, void *response) {
     p.writeInt32(p_cur->universal_pin_state);
     p.writeInt32(p_cur->gsm_umts_subscription_app_index);
     p.writeInt32(p_cur->cdma_subscription_app_index);
+    p.writeInt32(-1);
 
     sendSimStatusAppInfo(p, p_cur->num_applications, p_cur->applications);
 }
